flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -218,23 +221,127 @@ public void go() throws Exception {
+
+			casSinkFunc.enqueueResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+			Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+			casSinkFunc.invoke("hello");
+
+			Assert.assertEquals(0, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
+
+			completableFuture.complete(null);
+		}
+	}
+
+	@Test(timeout = 5000)
+	public void testReleaseOnSuccess() throws Exception {
+		final CassandraSinkBaseConfig config = CassandraSinkBaseConfig.newBuilder()
+			.setMaxConcurrentRequests(1)
+			.setMaxConcurrentRequestsTimeout(5, TimeUnit.MILLISECONDS)
+			.build();
+		try (TestCassandraSink casSinkFunc = new TestCassandraSink(config)) {
+			casSinkFunc.open(new Configuration());
+
+			CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+			casSinkFunc.enqueueResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+			Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+			casSinkFunc.invoke("hello");
+
+			Assert.assertEquals(0, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
+
+			completableFuture.complete(null);
+
+			Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+		}
+	}
+
+	@Test(timeout = 5000)
+	public void testReleaseOnFailure() throws Exception {
+		final CassandraSinkBaseConfig config = CassandraSinkBaseConfig.newBuilder()
+			.setMaxConcurrentRequests(1)
+			.setMaxConcurrentRequestsTimeout(5, TimeUnit.MILLISECONDS)
+			.build();
+		TestCassandraSink casSinkFunc = new TestCassandraSink(config);
+
+		casSinkFunc.open(new Configuration());
+
+		CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+		casSinkFunc.enqueueResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+		Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+		Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+		casSinkFunc.invoke("hello");
+
+		Assert.assertEquals(0, casSinkFunc.getAvailablePermits());
+		Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
+
+		completableFuture.completeExceptionally(new RuntimeException());
+
+		Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+		Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+		try {
+			casSinkFunc.close();
+		} catch (Exception e) {
+			Assert.assertTrue(e.getMessage().contains("Error while sending value"));
+		}
 	}
 
-	private static class TestCassandraSink extends CassandraSinkBase<String, ResultSet> {
+	@Test(timeout = 5000)
+	public void testTimeoutExceptionOnInvoke() throws Exception {
+		final CassandraSinkBaseConfig config = CassandraSinkBaseConfig.newBuilder()
+			.setMaxConcurrentRequests(0)
```

Comment:
what are the semantics for 0 maxConcurrentRequests?

Reply:
I used it for testing, but I now think it's a bad idea, so I'm going enforce that it must be greater than zero, and rewrite the test.